File: src/cpp/multithreading/13_thread_pool.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <future>
#include <chrono>
#include <random>
#include <deque>
#include <mutex>

int worker(int task_id) {
    std::cout << "Task " << task_id << " is starting..." << std::endl;
    std::random_device rd;
    std::mt19937 mt(rd());
    std::uniform_int_distribution<int> dist(1, 3);
    std::this_thread::sleep_for(std::chrono::seconds(dist(mt))); // Simulate some work
    int result = task_id * 2;
    std::cout << "Task " << task_id << " is finished. Result: " << result << std::endl;
    return result;
}

int main() {
    const int num_tasks = 10;
    const int num_workers = 3;
    std::vector<std::future<int>> futures;

    // Create a thread pool
    {
        std::vector<std::thread> threads;
        std::deque<int> task_ids;
        std::mutex task_ids_mutex;
        std::vector<std::packaged_task<int(int)>> tasks(num_tasks);

        // Initialize tasks and task_ids
        for (int i = 0; i < num_tasks; ++i) {
            tasks[i] = std::packaged_task<int(int)>(worker);
            task_ids.push_back(i);
        }

        // Worker function
        auto worker_fn = [&]() {
            while (true) {
                int task_id;
                {
                    std::lock_guard<std::mutex> lock(task_ids_mutex);
                    if (task_ids.empty()) {
                        break;
                    }
                    task_id = task_ids.front();
                    task_ids.pop_front();
                }
                tasks[task_id].operator()(task_id);
            }
        };

        // Launch worker threads
        for (int i = 0; i < num_workers; ++i) {
            threads.emplace_back(worker_fn);
        }

        // Join worker threads
        for (auto& thread : threads) {
            thread.join();
        }

        // Retrieve futures
        for (auto& task : tasks) {
            futures.push_back(task.get_future());
        }
    }

    // Collect the results as the tasks complete
    for (int i = 0; i < num_tasks; ++i) {
        int result = futures[i].get();
        std::cout << "Task " << i << " result collected: " << result << std::endl;
    }

    return 0;
}

File: src/python/multithreading/13_thread_pool.py
"""
This script demonstrates the use of concurrent.futures.ThreadPoolExecutor to create
and manage a thread pool, efficiently executing multiple tasks using a fixed number
of worker threads.

In this script, a fixed number of worker threads are created using ThreadPoolExecutor,
which manages the creation, execution, and completion of tasks. Each task is a function
(worker) that simulates some work and then returns a result. The tasks are submitted
to the ThreadPoolExecutor, which returns a Future object for each task. The Future
objects allow us to collect the results of the tasks as they complete, without the need
for manual thread management or synchronization.

Using ThreadPoolExecutor is an efficient way to manage a thread pool, as it takes care
of thread creation, work distribution, and resource management. Alternative approaches,
such as manually creating and managing threads, can be more error-prone and less efficient.

The main function in this script submits a fixed number of tasks to the ThreadPoolExecutor,
collects the results as the tasks complete, and then prints the results.
"""

import concurrent.futures
import random
import time


def worker(task_id):
    print(f"Task {task_id} is starting...")
    time.sleep(random.uniform(1, 3))  # Simulate some work
    result = task_id * 2
    print(f"Task {task_id} is finished. Result: {result}")
    return task_id, result


def main():
    num_tasks = 10
    num_workers = 3

    with concurrent.futures.ThreadPoolExecutor(max_workers=num_workers) as executor:
        # Submit tasks to the executor and store the resulting Future objects
        futures = [executor.submit(worker, i) for i in range(num_tasks)]

        # Collect the results as the tasks complete
        for future in concurrent.futures.as_completed(futures):
            task_id, result = future.result()
            print(f"Task {task_id} result collected: {result}")


if __name__ == "__main__":
    main()
